ir/visitor-crtp.h

```diff
@@ -0,0 +1,254 @@
+    void visitor_const_error() override;
+    bool check_clone(const Visitor *) override;
+
+    class ForwardChildren : public Visitor {
+        const ChangeTracker &visited;
+        const IR::Node *apply_visitor(const IR::Node *n, const char * = 0) {
+            if (visited.done(n)) return visited.result(n);
+            return n;
+        }
+
+    public:
+        explicit ForwardChildren(const ChangeTracker &v) : visited(v) {}
+    };
+ public:
+    profile_t init_apply(const IR::Node *root) override;
+    bool preorder(IR::Node *) { return true; }
+    void postorder(IR::Node *) {}
+    void revisit(const IR::Node *, const IR::Node *) {}
+    void loop_revisit(const IR::Node *) { BUG("IR loop detected"); }
+    void revisit_visited();
+    bool visit_in_progress(const IR::Node *) const;
+};
+
+template <class Derived>
+class ModifierCRTP : public ModifierCRTP_base {
+ public:
+    const IR::Node *apply_visitor(const IR::Node *n, const char *name = 0) override;
+    bool call_preorder(IR::Node *);
+    void call_postorder(IR::Node *);
+    void call_revisit(const IR::Node *, const IR::Node *);
+    void call_loop_revisit(const IR::Node *);
+};
+
+template <class Derived>
+const IR::Node *ModifierCRTP<Derived>::apply_visitor(const IR::Node *n, const char *name) {
+    if (ctxt) ctxt->child_name = name;
+    if (n) {
+        PushContext local(ctxt, n);
+        if (visited->busy(n)) {
+            call_loop_revisit(n);
+            // FIXME -- should have a way of updating the node?  Needs to be decided
+            // by the visitor somehow, but it is tough
+        } else if (visited->done(n)) {
+            call_revisit(n, visited->result(n));
+            n = visited->result(n);
+        } else {
+            visited->start(n, visitDagOnce);
+            IR::Node *copy = n->clone();
+            local.current.node = copy;
+            if (!dontForwardChildrenBeforePreorder) {
+                ForwardChildren forward_children(*visited);
+                copy->visit_children(forward_children);
+            }
+            visitCurrentOnce = visited->refVisitOnce(n);
+            if (call_preorder(copy)) {
+                copy->visit_children(*this);
+                visitCurrentOnce = visited->refVisitOnce(n);
+                call_postorder(copy);
+            }
+            if (visited->finish(n, copy)) (n = copy)->validate();
+        }
+    }
+    if (ctxt)
+        ctxt->child_index++;
+    else
+        visited.reset();
+    return n;
+}
+
+template <class Derived>
+bool ModifierCRTP<Derived>::call_preorder(IR::Node *n) {
+    switch (n->typeId()) {
+        //! TODO: Is it possible to generate only cases that are needed?
+        //        Can the optimizer get rid of unused cases, or that is too complicated?
+        //! TODO: Support Vector<*> and IndexedVector<*> nodes.
+#define DECLARE_NODE_CASE(CLASS, BASE)   \
+    case IR::NK_ ## CLASS: \
+        return static_cast<Derived*>(this)->preorder(n->checkedTo<IR::CLASS>());
```

Comment:
Actually, I think we'd just add `uncheckedTo`. As we already know the typeid, there is no need for extra check (`->to<T>` does typeinfo check).